Add suggestions component

opr opened the pull request:
### Changes proposed in this Pull Request:

This PR is based on #60955 

- Adds the `Suggestions` component to display autocomplete suggestions as the shopper types.
- Adds integration tests to ensure `AddressAutocomplete` and `Suggestions` work well together.

(For Bug Fixes) Bug introduced in PR # .

### Screenshots or screen recordings:

https://www.loom.com/share/03c42e37b38740f8806e839902564553

<!-- Begin testing instructions -->

### How to test the changes in this Pull Request:

<!-- Include detailed instructions on how these changes can be tested. Review and follow the guide for how to write high-quality testing instructions. -->

Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

1. Install WooPayments in test mode, or 
[mock-address-provider.zip](https://github.com/user-attachments/files/22388891/mock-address-provider.zip)

2. Go to WooCommerce -> Settings -> General and check `Enable predictive address search`
3. Go to the Checkout block, start typing into address 1, after 3 characters, address suggestions should show up.
4. Use the arrow keys and mouse to navigate between suggestions. Either click or press enter when one is highlighted and ensure the address gets entered into the form.
5. Test with screen reader, ensure switching between addresses in the suggestions component reads out correctly.

<!-- End testing instructions -->

### Testing that has already taken place:

<!-- Detail any testing that has already been conducted. -->
<!-- Include environment details such as hosting type, plugins, theme, store size, store age, and relevant settings. -->
<!-- Mention any analysis performed, such as assessing potential impacts on environment attributes and other plugins, performance profiling, or LLM/AI-based analysis. -->
<!-- Within the testing details you provide, please ensure that no sensitive information (such as API keys, passwords, user data, etc.) is included in this public pull request. -->

### Changelog entry

<!-- You can optionally choose to enter a changelog entry by checking the box below and supplying data. -->
<!-- It will trigger the 'Add changelog to PR' CI job to create and push the entry into the branch. -->

<!-- Due to org permissions, the job might fail for PRs crated from a fork under GitHub organizations. Possible solutions: -->
<!-- * Create entry manually with `pnpm --filter='@woocommerce/plugin-woocommerce' changelog add` and push it into the branch (replace `@woocommerce/plugin-woocommerce` with package name from nearest `package.json` file) -->
<!-- * Create entry from supplied PR data and push it automatically `pnpm utils changefile pr-number-here -o github-org-name-here` -->

-   [ ] Automatically create a changelog entry from the details below.

<!-- If no changelog entry is required for this PR, you can specify that below and provide a comment explaining why. This cannot be used if you selected the option to automatically create a changelog entry above. -->

-   [ ] This Pull Request does not require a changelog entry. (Comment required below)

<details>

<summary>Changelog Entry Details</summary>

#### Significance

<!-- Choose only one -->

-   [ ] Patch
-   [ ] Minor
-   [ ] Major

#### Type

<!-- Choose only one -->

-   [ ] Fix - Fixes an existing bug
-   [ ] Add - Adds functionality
-   [ ] Update - Update existing functionality
-   [ ] Dev - Development related task
-   [ ] Tweak - A minor adjustment to the codebase
-   [ ] Performance - Address performance issues
-   [ ] Enhancement - Improvement to existing functionality

#### Message <!-- Add a changelog message here -->

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment <!-- If your Pull Request doesn't require a changelog entry, a comment explaining why is required instead -->

</details>

opr committed:
Add suggestions component

opr committed:
Run search when user is searching

opr committed:
Forward inputRef to validated text input

opr committed:
Add address change handler

opr committed:
Add suggestions state var

opr committed:
Add suggestions component when search is finished

opr committed:
Copy suggestions CSS from shortcode

opr committed:
Import suggestions

opr committed:
Show provider branding at bottom of suggestions

opr committed:
Handle navigating suggestions with arrow keys

opr committed:
Handle clicking suggestions

opr committed:
Handle blurring the address input

opr committed:
Remove unneeded console error

opr committed:
Use mutation observer to disable autocomplete when searching

opr committed:
Show address 2 when filled by autocomplete

opr committed:
Add changelog

opr committed:
Disable a11y lint error

opr committed:
Add notice why debouncedSearch is not in deps array

opr committed:
Add standalone suggestions component tests

opr committed:
Add component/address1 integration test

opr committed:
Remove unused CSS

opr committed:
Update address form types

opr committed:
Remove isSearching

opr committed:
Disable autofill more reliably

opr committed:
Add mutation observer test

opr committed:
Fix lint issues

mikejolley reviewed (approved):
Followed steps and it did work. 

The only thing I noticed, non blocking, is that after typing a few characters and it doing the search, and on the first time I used the keyboard to navigate into the menu, there was a noticeable hang/delay of around a second before the next keystroke was registered. This was with the mock plugin.

opr commented:
> there was a noticeable hang/delay of around a second before the next keystroke was registered. This was with the mock plugin.

Thanks Mike, this is due to the implementation of the mock plugin making an AJAX request to get results, it could also be compounded by the debounce function, i'll look into perhaps cancelling in flight requests instead of debounce. May do this in a follow-up.

opr committed:
Remove debounce

opr committed:
Import autocomplete relatively to prevent cart store loading in test

opr commented:
@mikejolley I've removed the debounce now, I think providers should handle debouncing if they want to, we shouldn't really enforce it. The mock provider still delays by 300ms _and_ uses AJAX to fetch results.

I've rebased following changes upstream, so you may want to give this another spin before I merge. No changes here really, except cleanup following merge conflicts.

opr committed:
Harden checks accessing window props

opr committed:
Add optional chaining to prevent errors

tjcafferkey commented on plugins/woocommerce/client/blocks/assets/js/base/components/cart-checkout/address-autocomplete/address-autocomplete.tsx:
```diff
@@ -23,13 +34,355 @@ export const AddressAutocomplete = ( {
+					}
+				} )
+				.catch( () => {
+					setSuggestions( [] );
```
Non blocking: Maybe this is done on the server-side but when we are throwing/catching errors are we logging these anywhere for the merchant or us?

mikejolley reviewed (approved):
Gave a quick test after latest changes 👍🏻

opr committed:
Lint fix
